### PR TITLE
code coverage runtime plugin

### PR DESCRIPTION
**Pull request by brig**
Screenshot 2024-11-30 at 21 17 06

### CODE REVIEW

**Comment by brig** [comment]
docs: https://github.com/walmartlabs/concord-website/pull/177

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodecoverageModule.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.walmartlabs.concord.svm.ExecutionListener;
+
+import javax.inject.Named;
+
+@Named
+public class CodecoverageModule implements Module {
+
+    @Override
+    public void configure (Binder binder){
+        var taskProviders = Multibinder.newSetBinder(binder, ExecutionListener.class);
```
Comment (ibodrov) [inline runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodecoverageModule.java:35]:
```suggestion
        var executionListeners = Multibinder.newSetBinder(binder, ExecutionListener.class);
```

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        var fileNames = processDefinition.flows().values().stream()
+                .map(v -> v.location().fileName())
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+
+        persistenceService.persistFile(FLOWS_FILENAME, out -> {
+            try (ZipArchiveOutputStream zip = new ZipArchiveOutputStream(out)) {
+
+                for (var fileName : fileNames) {
+                    Path file = workDir.resolve(fileName);
```
Comment (ibodrov) [inline runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:163]:
```suggestion
                    var file = workDir.resolve(fileName);
```

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        return Result.CONTINUE;
+    }
+
+    @Override
+    public Result afterCommand(Runtime runtime, VM vm, State state, ThreadId threadId, Command cmd) {
+        if (cmd instanceof FlowCallCommand) {
+            processStep(((FlowCallCommand) cmd).getStep(), runtime, state, threadId);
+        }
```
Comment (ibodrov) [inline runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:90]:
```suggestion
        if (cmd instanceof FlowCallCommand fcc) {
            processStep(fcc.getStep(), runtime, state, threadId);
        }
```

Comment (ibodrov) [reply runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:90]: And similar stuff above.

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/StepInfo.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import javax.annotation.Nullable;
+
+@Value.Immutable
+@Value.Style(jdkOnly = true)
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonSerialize(as = ImmutableStepInfo.class)
+@JsonDeserialize(as = ImmutableStepInfo.class)
+public interface StepInfo {
```
Comment (ibodrov) [inline runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/StepInfo.java:37]: Maybe StepInfo can be a `record`?
The builder is used only in one place and can be replaced with a static builder method or a constructor.
Minus one dependency and a build step. :-)

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        generateReport(runtime);
+    }
+
+    private void generateReport(Runtime runtime) {
+        log.info("Generating code coverage info...");
+
+        try {
+            var reportProducer = new LcovReportProducer(runtime.getService(ProcessDefinition.class));
+
+            steps.stream().forEach(reportProducer::onStep);
```
Comment (ibodrov) [inline runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:130]: Just pass all steps to the producer?

**Review by ibodrov — COMMENTED** [review]
Some notes, didn't try running it yet.

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java (hunk 5 above)
Comment (brig) [reply runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:130]: hm... why not:)

File: runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java (hunk 3 above)
Comment (brig) [reply runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java]: :(
```
Error:  /home/runner/work/concord-plugins/concord-plugins/runtime/codecoverage/src/main/java/com/walmartlabs/concord/plugins/codecoverage/CodeCoverage.java:[88:44] Syntax error on token "fcc", delete this token
```

**Review by ibodrov — DISMISSED** [review]
A couple of minor comments, docs and the README need another pass. Otherwise LGTM.

**Review by ibodrov — APPROVED** [review]

**Pull request merged**